File: Leetcode/3 - Longest Substring Without Repeating Characters/solution.js
var lengthOfLongestSubstring = function (s) {
    let charCount = {};
    let left = 0, right = 0;
    let maxLength = 0;
    while (right < s.length) {
        const currChar = s[right];
        if (charCount[currChar]) {
            maxLength = Math.max(maxLength, (right - left));
            while (charCount[currChar]) {
                charCount[s[left]]--;
                left++;
            }
        } else {
            charCount[currChar] = (charCount[currChar] || 0) + 1;
            right++;
        }
    }
    return Math.max(maxLength, (right - left));
};

File: Leetcode/3 - Longest Substring Without Repeating Characters/solution.py
def lengthOfLongestSubstring(self, s: str) -> int:
    char_count = {}
    left = 0
    right = 0
    max_length = 0

    while right < len(s):
        curr_char = s[right]
        if char_count.get(curr_char, 0) > 0:
            max_length = max(max_length, right - left)
            while char_count.get(curr_char, 0) > 0:
                char_count[s[left]] -= 1
                left += 1
        else:
            char_count[curr_char] = char_count.get(curr_char, 0) + 1
            right += 1

    return max(max_length, right - left)